Add two new configuration options for periodic retained WAL size check in scaled down mode

alco opened the pull request:
This PR introduces periodic monitoring of retained WAL size when Electric has scaled down its database connections due to inactivity. If during one of these periodic checks the retained WAL is detected to have grown beyond the configured threshold, Electric wakes up the connection subsystem to resume replication stream processing.

## Core changes

`Connection.Restarter`
- When connections scale down, starts a periodic timer to check retained WAL size
- Queries Postgres for the retained WAL size using `pg_wal_lsn_diff(pg_current_wal_lsn(), restart_lsn)`
- If size exceeds threshold, restarts the connection subsystem
- Logs checks with human-readable sizes and intervals

`Connection.Manager`
When it validates connection options, it stores the validated ones in StackConfig. This can then be used by `Connection.Restarter` for one-off DB queries to check the retained WAL size. When `Connection.Manager` restarts (after an error or when the connection subsystem is woken up), it erases these options from StackConfig and repeats the validation process as before. This is to prevent the system from ending up in an invalid state by ensuring that at `Connection.Manager` startup it will always get the same starting connection options to work with.

Added two new configuration options:
- `ELECTRIC_REPLICATION_IDLE_WAL_SIZE_CHECK_PERIOD`: How often to check retained WAL size when scaled down (default: 1 hour)
- `ELECTRIC_REPLICATION_IDLE_WAL_SIZE_THRESHOLD`: WAL size threshold that triggers reconnection (default: 100 MiB)

Both options support human-readable time/size formats.

## Refactoring

Introduced a new module named `OneOffConnection` that wrap `Postgrex.SimpleConnection` and provides a simple API for opening a one-off DB connection, running a query (using the simple protocol) and getting the result back, all synchronously.

Reimplemented the lock breaking logic using `OneOffConnection` and removed `LockBreakerConnection` since it was no longer necessary to have as a separate module.

Refactored `ConnectionResolver`, replacing its ad-hoc wrapping of `Postgrex.SimpleConnection` with `OneOffConnection`.

## Testing

Added a new integration test (`integration-tests/tests/wal-size-check-while-scaled-down.lux`) that verifies Electric's handling of two cases during its periodic WAL size check: 1) WAL size under the threshold; 2) WAL size has exceeded the threshold.

---

Closes https://github.com/electric-sql/electric/issues/3260.

alco committed:
Store the result of connection opts validation in StackConfig

alco committed:
Implement periodic retained WAL size check in Restarter

alco committed:
Implement ELECTRIC_IDLE_WAL_SIZE_CHECK_PERIOD and ELECTRIC_IDLE_WAL_SIZE_THRESHOLD

alco committed:
Add changeset

msfstef reviewed (commented):
Looks good but I assume this does not work since there's no querying of the wal size (and no tests which I assume is for the above reason?)

alco committed:
Add Electric.Postgres.OneOffConnection module to run one-off DB queries

alco committed:
Reimplement Electric.Postgres.LockBreakerConnection as a one-off query

alco committed:
Remove Electric.Connection.Manager.ConnectionResolver's internal SimpleConnection and use OneOffConnection

alco committed:
Query retained WAL size in Restarter periodically

alco committed:
Split OneOffConnection's API into separate attempt_connection() and query() functions

alco committed:
Pass the user config for periodic WAL size checks to Restarter

alco committed:
Pretty-print retained WAL size and threshold when logging Restarter notices

alco committed:
Set a more sensible default for WAL size threshold

alco committed:
Remove superfluous logging from OneOffConnection

alco committed:
Do not persist validation connection opts across conn man restarts

To avoid the failure state where Electric can no longer connect to the
db without first doing the revalidation

alco committed:
fixup! Do not persist validation connection opts across conn man restarts

alco committed:
Use consistent prefix ELECTRIC_REPLICATION_IDLE for the new configuration opts

alco committed:
Cleanups for the new Lux test

alco committed:
Parse WAL size threshold as human-readable size value

msfstef commented on packages/sync-service/lib/electric/connection/restarter.ex:
```diff
@@ -10,12 +10,18 @@ defmodule Electric.Connection.Restarter do
     - schema reconciler
 
+  Once the connection subsystem is scaled down, Restarter starts a timer to check the retained
+  WAL size periodically. If the size exceeds the configured threshold, Restart will restart the
```
```suggestion
  WAL size periodically. If the size exceeds the configured threshold, Restarter will restart the
```

msfstef commented on packages/sync-service/lib/electric/connection/restarter.ex:
```diff
@@ -108,11 +126,92 @@ defmodule Electric.Connection.Restarter do
+    {:noreply, state}
+  end
+
+  def handle_info(:check_wal_size, state) do
```
I've seen the integration test - would also appreciate a Restarter only integration test as part of our elixir test suite if it's not too difficult to set up!

msfstef reviewed (approved):
Love the one off connection utility!

Looks great, would appreciate a couple of happy path elixir tests isolating the restarter
